Add the option to create Changelogs grouped by Build to generate_changelog.py

KurzedMetal opened the pull request:
#### Summary
```Summary: Infrastructure "Add the option to create Changelogs grouped by Build to generate_changelog.py"```

#### Purpose of change
To have a different way to show changelogs to users.
This new output is a strictly improved version of the Changelog generated by Jenkins, which is the one used by the CDDA Launcher and only lists commit messages instead of PR "Summaries".
I think both CDDA developers and the Launcher developers could be interested in this extra Jenkins info.

#### Describe the solution
I added extra classes and command-line arguments to the previously submitted script in PR #25840.
Now the code is able to access Jenkins API to get information about Builds and relate Builds to Commits and finally to PRs.
It can still generate exactly the same output as before with the argument `--group-by-date` and the new output is generated with `--group-by-build`. One of these two arguments must be picked or the script will show an error.

Example of the new output:
![image](https://user-images.githubusercontent.com/1426680/46451877-27ede700-c76f-11e8-831e-4e88262a173a.png)

KurzedMetal commented:
BTW, do you guys think I should change the outout of --`group-by-build` to something similar as for `--group-by-date`?

Like:
![image](https://user-images.githubusercontent.com/1426680/46483280-a37e8100-c7cd-11e8-8347-0a193ece8080.png)

Instead of:
![image](https://user-images.githubusercontent.com/1426680/46483291-aaa58f00-c7cd-11e8-894b-8c57b909617e.png)

I think I initially avoid it because sometimes some builds have very few lines or just one of each type of "Summary", instead of multiple for each category.

Anyways, any of this is easily modifiable.

KurzedMetal commented:
After more coding, tests and some refactor/bugfixing, I don't think I can improve the logic for matching Builds with PRs/Commits more than currently is. This is the best it will get (which is quite good).

So far I've found the only offenders are merge commits that have messages like: `Merge remote-tracking branch 'origin/pr/25410'`. These kind of merge are screwing both Merge Hash and Commit / PR Merge dates enough to give me a hard time on matching the PR vs the Commit. The only thing I have left to do is probably parsing the messages (if they don't change much), which I was avoiding, but I guess there's no option left.

As for the output layout/grouping, if anyone has any recommendation, I'm all ears.

AMurkin commented:
That's not mine I think.
```
2018-09-25
    MISC. COMMITS
        * Merge remote-tracking branch 'origin/pr/25191' (by AMurkin in Commit 2e596ef)
```

KurzedMetal commented:
Thanks for pointing that out, I noticed that earlier and fixed it, but I fixed it for "by build" and not "by date" and then it crept back to "by build" when I made them both look alike.
